src/www/app/app.ts

```diff
@@ -543,10 +548,16 @@ export class App {
     return server;
   }
 
-  // Returns the card associated with serverId, throws if no such card exists.
-  // See server-list.html.
-  private getCardByServerId(serverId: string) {
-    return this.serverListEl.getServerCard(serverId);
+  private updateUiServer(id: string, properties: object) {
+    // We have to create a new list so the property change is observed.
```

Comment:
Do you feel that maybe this commitment to building fresh objects is the concern of some kind of Repository?

Reply:
"Repository" is a DDD concept and is part of the domain model. It's not aware of the UI (think of the model in Model-View-Presenter, but at a higher level)

Connecting the domain model and the UI is the job of the application code (the "Presenter" in this analogy).

Note that there's actual MVP happening here at a lower level. The `ServerCardModel[]` is part of the model of the ServerList. The ServerList html is part of the "View", and the implementation of the component the "Presenter".

---
See Page 17 at https://www.domainlanguage.com/wp-content/uploads/2016/05/DDD_Reference_2015-03.pdf for more on Repositories.

This might be helpful too: https://dzone.com/refcardz/getting-started-domain-driven

I think they are good resources, but I've tweaked a bit our use of DDD.

Reply:
That's mixing the layers. The Domain Model is a very particular thing. It represents domain concepts exposed to the user. The UI card model is an implementation detail. They are at different abstraction layers.

Repositories are for creating, persisting, retrieving and destroying Domain Model Entities. The UI model is not an Entity. It doesn't have an identity and it doesn't need to be persisted. It reflects aa snapshot of the state of an Entity.

The application code is responsible for updating the UI in response to changes to the Domain Model. You can do that based on events or the results of operations.

Reply:
Ah, alright. I suppose if this specifically becomes a recurring issue we can just use a `deepClone` or something like that. Thanks for the explanation.

Reply:
The way the rendering triggers is by checking if the property changed value. It's a simple equality test. A deep clone is not needed. Changes within the object don't trigger a re-render.

Reply:
If it's a list of objects where each object maps to a sub-component, we'd need to clone both the list and the sub-objects, which a `deepClone` would handle. That said, it's overkill, forget i said anything